treeherder/perf/models.py

```diff
@@ -107,32 +108,94 @@ class PerformanceDatumManager(models.Manager):
+        elapsed_runtime = now - started_at
+
+        return max_overall_runtime < elapsed_runtime
+
+    def cycle_data(self, repository, cycle_interval, chunk_size, sleep_time,
+                   unsheriffed_repos_expire_days, logger,
+                   started_at, max_overall_runtime):
         """Delete data older than cycle_interval, splitting the target data
 into chunks of chunk_size size."""
-
-        max_timestamp = datetime.datetime.now() - cycle_interval
-
-        # seperate datums into chunks
-        while True:
-            perf_datums_to_cycle = list(self.filter(
-                repository=repository,
-                push_timestamp__lt=max_timestamp).values_list('id', flat=True)[:chunk_size])
-            if not perf_datums_to_cycle:
-                # we're done!
-                break
-            self.filter(id__in=perf_datums_to_cycle).delete()
-            if sleep_time:
-                # Allow some time for other queries to get through
-                time.sleep(sleep_time)
+        logger.warning('Cycling Perfherder data from {0} repository...'
+                       .format(repository.name))
+        max_timestamp = self._determine_max_timestamp(repository.name,
+                                                      cycle_interval,
+                                                      unsheriffed_repos_expire_days)
+
+        try:
+            self._delete_in_chunks(repository, max_timestamp, chunk_size, sleep_time, logger,
+                                   started_at, max_overall_runtime)
+        except NoDataCyclingAtAll as ex:
+            logger.warning('Exception: {}'.format(ex))
+            return
 
         # also remove any signatures which are (no longer) associated with
         # a job
+        logger.warning('Removing performance signatures with missing jobs...')
         for signature in PerformanceSignature.objects.filter(
                 repository=repository):
-            if not self.filter(signature=signature).exists():
+            if not self.filter(
+                    repository=repository,  # redundant, but leverages (repository, signature) compound index
+                    signature=signature).exists():
                 signature.delete()
 
+    def _delete_in_chunks(self, repository, max_timestamp, chunk_size, sleep_time, logger,
+                          started_at, max_overall_runtime):
+        from django.db import connection
+        any_succesful_attempt = False
+
+        with connection.cursor() as cursor:
+            while True:
+                if self._should_quit(started_at, max_overall_runtime):
+                    logger.info('Max runtime exceeded. Quitting cycling for repository {}'
+                                .format(repository.name))
+                    break
+
+                try:
+                    cursor.execute('''
+                        DELETE FROM `performance_datum`
+                        WHERE repository_id = %s AND
```

Comment:
Does the `repository_id` matter?  Can all repos be done at once?  If so, this can run faster, and leave less holes in the table.

I am also concerned that a full table scan is required if the `chunk_size > number_of_records_to_delete`, which will then timeout.  Maybe a rough upper bound on the `id`can help with that:
    
    select id into @max_id from performance_datum where push_timestamp>'2018-07-22' limit 1;
    select id from performance_datum where push_timestamp<'2018-07-22' and id<@max_id limit 1000;

Reply:
> Does the `repository_id` matter? Can all repos be done at once? If so, this can run faster, and leave less holes in the table.

Not really, we could delete rows without it.

> I am also concerned that a full table scan is required if the `chunk_size > number_of_records_to_delete`, which will then timeout. Maybe a rough upper bound on the `id`can help with that:
> 
> ```
> select id into @max_id from performance_datum where push_timestamp>'2018-07-22' limit 1;
> select id from performance_datum where push_timestamp<'2018-07-22' and id<@max_id limit 1000;
> ```

@klahnakoski I guess it's best to compute this upper bound right before running the DELETE ... LIMIT query.